### PR TITLE
Added MDP generation to QEff Compile

### PR DESCRIPTION
**Pull request by quic-mohmeh**
This PR adds the MDP generation required in case of disaggregated serving for Prefill. This supports both Pipeline Prefill + Tensor Slicing and also supports passing custom cores to the MDP generator

### CODE REVIEW

**Comment by quic-mohmeh** [comment]
Tested and working on the following model classes

- CodeLlama-7b-Instruct
- falcon-7b-instruct
- gemma-2-9b-it
- gpt-oss-20b
- granite-3.1-8b-instruct
- Llama-3.2-1B-Instruct
- Llama-3.2-3B
- Phi-3-mini-4k-instruct

**Comment by quic-rishinr** [comment]
@mamtsing @ochougul please review the PR

**Comment by quic-mohmeh** [comment]
@quic-rishinr @mamtsing @ochougul  A gentle reminder for review

File: QEfficient/base/modeling_qeff.py
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py:1]: Add warning that ignores `mdp_ts_num_partitions` whenever `seq_len==1`
Also, add a warning that ignores when `ts_num_devices>1` and `seq_len>` and `mdp_ts_num_partitions>1`

**Comment by quic-mohmeh** [comment]
To be tested on:
- Qwen3 Dense
- Qwen3 MOE

- Test with subfunctions

**Comment by quic-hemagnih** [comment]
@quic-mohmeh Please let us know once your testing is complete. Also incorporate the review comments. @quic-mamta Can also please review it.

**Comment by quic-mohmeh** [comment]
Works on Qwen3-4B(Dense) - without subfunctions

**Comment by quic-mohmeh** [comment]
Works on Qwen3-30B-A3B (MOE) model as well - without subfunctions

File: QEfficient/compile/mdp_generator.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    # Registered with DEFINEKNOWNCUSTOMOP in ONNXModelLoader.cpp
+    _KNOWN_CUSTOM_OPS = frozenset({"CustomRMSNorm"})
+
+    local_functions = {f.name: f for f in model.functions}
+    logger.info(f"Found {len(local_functions)} local function types: {set(local_functions.keys())}")
+
+    inlined_funcs: Set[str] = set()
+    non_inlined_funcs: Set[str] = set()
+    for func_name, func in local_functions.items():
+        if func_name in _KNOWN_CUSTOM_OPS or len(func.node) >= 100:
```
Comment (quic-mamta) [inline QEfficient/compile/mdp_generator.py:119]: why do we have condition for < 100 nodes here? is this tested with subfunctions?

**Comment by quic-mohmeh** [comment]
Update example scripts for Qwen3(with and without VL) and GPTOSS

**Comment by quic-rishinr** [comment]
@quic-mohmeh please rebase it on top of release/v1.22.0_tmp branch

**Comment by quic-mohmeh** [comment]
Tested with subfunctions on the following model:
- CodeLlama-7b-Instruct-hf
- falcon-7b-instruct
- gemma-2-9b-it
- granite-3.1-8b-instruct
- Llama-3.1-70B-Instruct
- Llama-3.2-1B-Instruct
- Llama-3.2-3B
- Phi-3-mini-4k-instruct
- Qwen3-30B-A3B
- GPTOSS-20B

File: QEfficient/base/modeling_qeff.py [hunk 2]
```diff
@@ -516,6 +516,7 @@ def _compile(
         specializations: Optional[List[Dict[str, int]]] = None,
         custom_io: Optional[Dict[str, str]] = None,
         mdp_ts_num_devices: int = 1,
+        mdp_num_partitions: int = 1,
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py]: nit: add it as an optional argument

File: QEfficient/compile/mdp_generator.py (hunk 1 above)
Comment (quic-mohmeh) [reply QEfficient/compile/mdp_generator.py:119]: It is there in compiler code, so I have adapted the same and yes, the subfunctions are supported

**Comment by quic-mohmeh** [comment]
Also verified by Qeff team on following models with subfunctions:
- Qwen/Qwen3-235B-A22B (Karthikeya)
- moonshotai/Kimi-K2-Instruct (Mamta)

File: QEfficient/compile/mdp_generator.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        onnx_path:      Path to the exported ONNX file.
+        num_devices:    Total devices (num_partitions * ts_per_stage).
+        num_partitions: Number of pipeline stages.
+        num_layers:     Number of transformer layers.
+        num_cores:      NSP cores per device (default 16).
+
+    Returns:
+        dict with keys 'connections' and 'partitions'.
+    """
+    assert num_partitions <= num_devices, f"num_partitions ({num_partitions}) must be <= num_devices ({num_devices})"
```
Comment (quic-rishinr) [inline QEfficient/compile/mdp_generator.py]: instead of assert pease raise an exection and add a check for num_partitions > 0 for avoiding the zero division error

File: QEfficient/base/modeling_qeff.py [hunk 4]
```diff
@@ -602,22 +608,48 @@ def _compile(
...
-        elif mdp_ts_json_path:
+        if mdp_ts_json_path:
             command.append(f"-mdp-load-partition-config={mdp_ts_json_path}")
             mdp_ts_json = load_json(str(mdp_ts_json_path))
+        elif mdp_num_partitions > 1 or "stages" in compiler_options:
+            mdp_num_partitions = compiler_options.pop("stages", mdp_num_partitions)
+            # Disaggregated (pipeline-parallel) MDP: generate a fully-populated
+            # nodeList per partition directly from the ONNX graph — no compiler
+            # round-trip required.
+            from QEfficient.compile.mdp_generator import generate_disagg_mdp_partition_config
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py]: nit: please move the import at the top

File: QEfficient/base/modeling_qeff.py [hunk 5]
```diff
@@ -602,22 +608,48 @@ def _compile(
...
-        # MDP partition config: prioritize dump over load
-        mdp_dump_json_path = compiler_options.pop("mdp_dump_partition_config", None)
+        # MDP partition config selection (three priorities, highest first):
+        #   1. User explicitly provides a pre-built MDP JSON to load.
+        #   2. Disaggregated (pipeline-parallel) MDP — generate from ONNX topsort.
+        #   3. Template (tensor-slice) MDP — single partition, nodeList absent.
         mdp_ts_json_path = compiler_options.pop("mdp_load_partition_config", None)
+        # Silently discard any stale mdp_dump_partition_config key that callers
+        # may still pass; the compiler-round-trip dump path is no longer supported.
+        compiler_options.pop("mdp_dump_partition_config", None)
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py]: if user want to dumpt the partition config would we need to have to capture the flag and dumpt the flag? also we should not be silently poping out any flag. we should be providing proper warnings

File: QEfficient/base/modeling_qeff.py [hunk 6]
```diff
@@ -602,22 +608,48 @@ def _compile(
...
-            if mdp_ts_json_path:
-                logger.warning(
-                    "Loading and Dumping partition is not supported at the same time. Prioritizing dump config over load config!"
-                )
-            command.append(f"-mdp-dump-partition-config={mdp_dump_json_path}")
-        elif mdp_ts_json_path:
+        if mdp_ts_json_path:
             command.append(f"-mdp-load-partition-config={mdp_ts_json_path}")
             mdp_ts_json = load_json(str(mdp_ts_json_path))
+        elif mdp_num_partitions > 1 or "stages" in compiler_options:
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py]: What is this stages kwarg we are handling? is it a compiler flag?

File: QEfficient/compile/mdp_generator.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
+# -----------------------------------------------------------------------------
```
Comment (quic-rishinr) [inline QEfficient/compile/mdp_generator.py:1]: please add unit tests and test for mdp generation. Please make sure the tests are small and exectue under 5 to 10 seconds.

File: QEfficient/compile/mdp_generator.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    Returns:
+        inlined_node_map:   dict mapping call-site name -> list of inlined
+                            sub-node names (<call_site>/<func_node>).
+        non_inlined_funcs:  set of function names that are NOT inlined
+                            (known custom ops or >= 100 nodes); their
+                            call-site names are valid nodeList entries.
+    """
+    # Op types registered via DEFINEKNOWNCUSTOMOP in ONNXModelLoaderCustomOp.cpp.
+    # The compiler loads these as a single named node (not expanded/inlined).
+    _KNOWN_CUSTOM_OPS = frozenset(
```
Comment (quic-rishinr) [inline QEfficient/compile/mdp_generator.py:106]: Can we fetch it from pytorch transforms? than kepping 2 copies?

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

**Review by quic-akuruvil — COMMENTED** [review]
@quic-mohmeh Have you verified PP+TS combination with this? Or only PP works with this? WHen you say you have verified the above models, have you verified the output sanity too?

**Comment by quic-mohmeh** [comment]
> @quic-mohmeh Have you verified PP+TS combination with this? Or only PP works with this? WHen you say you have verified the above models, have you verified the output sanity too?

As vLLM currently doesn't support PP+TS, so I am unable to verify PP+TS. As for output validation, I have only skim the output, the main purpose of these tests is to check whether the model compiles or not with the MDP generated

File: QEfficient/base/modeling_qeff.py (hunk 2 above)
Comment (quic-mohmeh) [reply QEfficient/base/modeling_qeff.py]: Done

File: QEfficient/base/modeling_qeff.py (hunk 5 above)
Comment (quic-mohmeh) [reply QEfficient/base/modeling_qeff.py]: Thanks for pointing this out. Yes, in case of old compiler flow, user want to dump the MDP first from the compiler

File: QEfficient/base/modeling_qeff.py (hunk 6 above)
Comment (quic-mohmeh) [reply QEfficient/base/modeling_qeff.py]: No it is a vLLM specific flag, but as we discussed to remove the redundancy, I am removing the `stages` argument in favor of single argument - `mdp_num_partitions`

File: QEfficient/base/modeling_qeff.py (hunk 4 above)
Comment (quic-mohmeh) [reply QEfficient/base/modeling_qeff.py]: Done

File: QEfficient/compile/mdp_generator.py (hunk 7 above)
Comment (quic-mohmeh) [reply QEfficient/compile/mdp_generator.py:1]: @quic-rishinr as we have discussed that there is no valid approach to compare compiler MDP dump and QEff MDP dump, all we can ensure is that every node present in compiler MDP dump should be present in QEff MDP dump in correct order. For verifying and testing, we might need to compile no PP and PP QPCs and compare the output. Let me know which direction you want to proceed for testing

File: QEfficient/compile/mdp_generator.py (hunk 8 above)
Comment (quic-mohmeh) [reply QEfficient/compile/mdp_generator.py:106]: These operators are retrieved from Compiler codebase and are part of optimizations compiler need before dumping the MDP. This differs from pytorch transforms

File: QEfficient/compile/mdp_generator.py (hunk 3 above)
Comment (quic-mohmeh) [reply QEfficient/compile/mdp_generator.py]: Done

**Review by quic-mohmeh — COMMENTED** [review]
Addressed the comments @quic-rishinr

**Comment by quic-mohmeh** [comment]
@quic-rishinr @quic-hemagnih 
Needs to be tested with VLMs as well, please do not merge yet.

**Comment by quic-mohmeh** [comment]
VLMs are not working currently with this MDP partition, both with and without subfunctions(E-P-D). Have to look into compiler codebase

**Comment by quic-akuruvil** [comment]
@quic-mohmeh Can you rebase your PR with latest

**Comment by quic-mohmeh** [comment]
Working on VLMs - tested on Qwen2.5VL 3B with and without subfunctions(PP4)

**Comment by quic-akuruvil** [comment]
> Working on VLMs - tested on Qwen2.5VL 3B with and without subfunctions(PP4)

Hi @quic-mohmeh When you say tested on VLMS, have you enabled PP on prefill part of language model? Or have you checked with PP on vision part as well?

**Comment by quic-mohmeh** [comment]
> > Working on VLMs - tested on Qwen2.5VL 3B with and without subfunctions(PP4)
> 
> Hi @quic-mohmeh When you say tested on VLMS, have you enabled PP on prefill part of language model? Or have you checked with PP on vision part as well?

Yes, PP with Prefill, TS on Encode and Decode

**Comment by quic-vargupt** [comment]
Gemma4-26B-A4B Prefill compiled with subfunctions

File: QEfficient/base/modeling_qeff.py [hunk 9]
```diff
@@ -793,22 +800,47 @@ def _compile(
...
-        elif mdp_ts_json_path:
+        if mdp_ts_json_path:
             command.append(f"-mdp-load-partition-config={mdp_ts_json_path}")
             mdp_ts_json = load_json(str(mdp_ts_json_path))
+        elif mdp_num_partitions > 1:
+            # Disaggregated (pipeline-parallel) MDP: generate a fully-populated
+            # nodeList per partition directly from the ONNX graph — no compiler
+            # round-trip required.
+
+            num_layers = getattr(self, "num_layers", None)
```
Comment (quic-akuruvil) [inline QEfficient/base/modeling_qeff.py:920]: @quic-mohmeh Are you still going with `stages` in user API? If so, where is it popped before qaic-compile, since it is not a valid compiler option?

File: QEfficient/compile/mdp_generator.py (hunk 7 above)
Comment (quic-akuruvil) [reply QEfficient/compile/mdp_generator.py:1]: Hi @quic-mohmeh , for unit test case's part, one of the required cases is, add a e2e tiny 2L model output matching verification. This model should be compiled with the manual MDP you have prepared. You can refer and re-use some of the test APIs we already use for output matching against reference. This should test the functionality of the change.

**Review by quic-akuruvil — CHANGES_REQUESTED** [review]

File: QEfficient/compile/mdp_generator.py (hunk 7 above)
Comment (quic-akuruvil) [reply QEfficient/compile/mdp_generator.py:1]: @quic-mohmeh Add add other unit test cases to flag any edge cases and so on, which causes the functionality to break.

**Comment by quic-mohmeh** [comment]
@quic-rishinr As QEff has moved from tmp branch to mdp branch this PR is no longer relevant. Created new PR: https://github.com/quic/efficient-transformers/pull/1086
